python/ml4ir/base/model/metrics/metrics_impl.py

```diff
@@ -0,0 +1,75 @@
+import tensorflow as tf
+from tensorflow.keras import metrics
+
+from typing import List
+
+
+class SegmentMean(metrics.Metric):
+    """
+    Keras Metrics class that computes a mean of the segments instead of overall mean.
+
+    Can be used to compute macro, micro averages by overriding result()
+    """
+
+    def __init__(self, segments: List[str] = [], name: str = "segment_mean", **kwargs):
+        """
+        Instantiate a SegmentMean object
+
+        Parameters
+        ----------
+        num_segments: int
+            Number of segments to be used for computing the mean
+        name: str
+            Name of the metric
+        **kwargs: dict
+            Additional keyword arguments to be passed to the metric constructor
+        """
+        super().__init__(name=name, **kwargs)
+        if not segments:
+            raise ValueError(f"Invalid argument passed for segments -> {segments}")
+        self.segments = segments
+        # NOTE: number of segments is vocabulary + 1 for OOV bucket
+        self.num_segments = tf.constant(len(segments) + 1)
+        self.segment_lookup = tf.keras.layers.StringLookup(vocabulary=segments,
+                                                           num_oov_indices=1)
+        self.total_sum = self.add_weight(name="total_sum", shape=(self.num_segments,),
+                                         initializer="zeros")
+        self.total_count = self.add_weight(name="total_count", shape=(self.num_segments,),
+                                           initializer="zeros")
```

Comment:
@mohazahran check this section for how segment metrics work. Basically, 
1. have a mapping table for string to ID
2. instead of scalar total and counts, use vector where each ID from step 1 is mapped to index of tensor